File: frontend/src/api/budgets.ts
/**
 * Budget Management API Client
 *
 * Handles all budget-related API calls for user budget tracking and management.
 */

import axios from 'axios';
import { API_ENDPOINTS } from '../utils/constants';

export interface UserBudget {
  id: string;
  user_id: string;
  monthly_budget_aud: number;
  current_spend_aud: number;
  budget_period_year: number;
  budget_period_month: number;
  alert_threshold_percent: number;
  is_active: boolean;
  set_by_user_id?: string;
  created_at: string;
  updated_at: string;
  budget_remaining: number;
  budget_utilization_percent: number;
  is_over_budget: boolean;
  should_alert: boolean;
}

export interface UserBudgetCreate {
  user_id: string;
  monthly_budget_aud: number;
  budget_period_year: number;
  budget_period_month: number;
  alert_threshold_percent?: number;
}

export interface UserBudgetUpdate {
  monthly_budget_aud?: number;
  alert_threshold_percent?: number;
  is_active?: boolean;
}

export interface TeamMemberBudget {
  user_id: string;
  user_name?: string;
  user_email: string;
  monthly_budget_aud: number;
  current_spend_aud: number;
  budget_remaining: number;
  budget_utilization_percent: number;
  is_over_budget: boolean;
  should_alert: boolean;
  budget_period_year: number;
  budget_period_month: number;
}

export interface TeamBudgetOverview {
  total_team_budget: number;
  total_team_spend: number;
  total_team_remaining: number;
  average_utilization_percent: number;
  users_over_budget: number;
  users_near_threshold: number;
  team_members: TeamMemberBudget[];
}

/**
 * Get current user's budget
 */
export async function getMyBudget(
  year?: number,
  month?: number
): Promise<UserBudget> {
  const params = new URLSearchParams();
  if (year) params.append('year', year.toString());
  if (month) params.append('month', month.toString());

  const url = `${API_ENDPOINTS.BUDGETS.ME}${params.toString() ? '?' + params.toString() : ''}`;
  const response = await axios.get<UserBudget>(url);
  return response.data;
}

/**
 * Get a specific user's budget (managers only)
 */
export async function getUserBudget(
  userId: string,
  year?: number,
  month?: number
): Promise<UserBudget> {
  const params = new URLSearchParams();
  if (year) params.append('year', year.toString());
  if (month) params.append('month', month.toString());

  const url = `${API_ENDPOINTS.BUDGETS.USER(userId)}${params.toString() ? '?' + params.toString() : ''}`;
  const response = await axios.get<UserBudget>(url);
  return response.data;
}

/**
 * Set or update a user's budget (managers only)
 */
export async function setUserBudget(
  userId: string,
  budgetData: UserBudgetCreate
): Promise<UserBudget> {
  const response = await axios.post<UserBudget>(
    API_ENDPOINTS.BUDGETS.USER(userId),
    budgetData
  );
  return response.data;
}

/**
 * Update an existing budget (managers only)
 */
export async function updateBudget(
  budgetId: string,
  budgetData: UserBudgetUpdate
): Promise<UserBudget> {
  const response = await axios.patch<UserBudget>(
    API_ENDPOINTS.BUDGETS.UPDATE(budgetId),
    budgetData
  );
  return response.data;
}

/**
 * Get team budget overview (managers only)
 */
export async function getTeamBudgetOverview(
  year?: number,
  month?: number
): Promise<TeamBudgetOverview> {
  const params = new URLSearchParams();
  if (year) params.append('year', year.toString());
  if (month) params.append('month', month.toString());

  const url = `${API_ENDPOINTS.BUDGETS.TEAM_OVERVIEW}${params.toString() ? '?' + params.toString() : ''}`;
  const response = await axios.get<TeamBudgetOverview>(url);
  return response.data;
}

File: backend/app/api/v1/endpoints/budgets.py
"""
Budget Management API endpoints.

Endpoints for managers to set user budgets and for users to track their spending.
"""

from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.base import get_db
from app.models import User
from app.schemas.user_budget import (
    UserBudgetCreate,
    UserBudgetUpdate,
    UserBudgetResponse,
    UserBudgetWithUser,
    TeamBudgetOverview,
)
from app.services.budget_service import BudgetService

# Note: Authentication dependencies would be added here
# For now, using placeholder dependency
async def get_current_user() -> User:
    """Placeholder for authentication dependency."""
    # TODO: Implement actual authentication
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="Authentication not implemented yet",
    )


router = APIRouter(prefix="/budgets", tags=["budgets"])


@router.post(
    "/users/{user_id}",
    response_model=UserBudgetResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Set user budget",
    description="Set or update a user's monthly budget (managers only)",
)
async def set_user_budget(
    user_id: str,
    budget_data: UserBudgetCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Set or update a user's monthly budget.

    **Permissions**: Department managers can set budgets for users in their department.
    Org admins and super admins can set budgets for any user.

    Args:
        user_id: Target user ID
        budget_data: Budget data (amount, period, alert threshold)
        db: Database session
        current_user: Authenticated user (must be manager)

    Returns:
        UserBudgetResponse: Created or updated budget

    Raises:
        403: If user is not a manager
        404: If target user not found
    """
    # Check if current user is a manager
    if not current_user.is_manager:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only managers can set user budgets",
        )

    # Override user_id in budget_data to match path parameter
    budget_data.user_id = user_id

    # Set the budget
    budget = await BudgetService.set_user_budget(
        db=db,
        budget_data=budget_data,
        set_by_user_id=current_user.id,
    )

    return budget


@router.get(
    "/me",
    response_model=UserBudgetResponse,
    summary="Get my budget",
    description="Get current user's budget for current month",
)
async def get_my_budget(
    year: Optional[int] = None,
    month: Optional[int] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get the current user's budget.

    Args:
        year: Budget year (optional, defaults to current year)
        month: Budget month (optional, defaults to current month)
        db: Database session
        current_user: Authenticated user

    Returns:
        UserBudgetResponse: User's budget information
    """
    budget = await BudgetService.get_user_budget(
        db=db,
        user_id=current_user.id,
        year=year,
        month=month,
    )

    if not budget:
        # Create default budget if none exists
        budget = await BudgetService.get_or_create_current_budget(
            db=db,
            user_id=current_user.id,
        )

    return budget


@router.get(
    "/users/{user_id}",
    response_model=UserBudgetResponse,
    summary="Get user budget",
    description="Get a specific user's budget (managers only)",
)
async def get_user_budget(
    user_id: str,
    year: Optional[int] = None,
    month: Optional[int] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get a user's budget (managers only).

    Args:
        user_id: Target user ID
        year: Budget year (optional)
        month: Budget month (optional)
        db: Database session
        current_user: Authenticated user (must be manager)

    Returns:
        UserBudgetResponse: User's budget information

    Raises:
        403: If current user is not a manager
        404: If budget not found
    """
    # Check permissions
    if not current_user.is_manager:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only managers can view other users' budgets",
        )

    budget = await BudgetService.get_user_budget(
        db=db,
        user_id=user_id,
        year=year,
        month=month,
    )

    if not budget:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Budget not found for this user and period",
        )

    return budget


@router.patch(
    "/{budget_id}",
    response_model=UserBudgetResponse,
    summary="Update budget",
    description="Update an existing budget (managers only)",
)
async def update_budget(
    budget_id: str,
    budget_data: UserBudgetUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Update an existing budget.

    Args:
        budget_id: Budget ID to update
        budget_data: Update data
        db: Database session
        current_user: Authenticated user (must be manager)

    Returns:
        UserBudgetResponse: Updated budget

    Raises:
        403: If user is not a manager
        404: If budget not found
    """
    if not current_user.is_manager:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only managers can update budgets",
        )

    budget = await BudgetService.update_user_budget(
        db=db,
        budget_id=budget_id,
        budget_data=budget_data,
    )

    if not budget:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Budget not found",
        )

    return budget


@router.get(
    "/team/overview",
    response_model=TeamBudgetOverview,
    summary="Get team budget overview",
    description="Get budget overview for all team members (managers only)",
)
async def get_team_budget_overview(
    year: Optional[int] = None,
    month: Optional[int] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Get budget overview for all users in manager's department.

    Args:
        year: Budget year (optional, defaults to current year)
        month: Budget month (optional, defaults to current month)
        db: Database session
        current_user: Authenticated user (must be manager)

    Returns:
        TeamBudgetOverview: Team budget statistics and individual budgets

    Raises:
        403: If user is not a manager
    """
    if not current_user.is_manager:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only managers can view team budget overview",
        )

    overview = await BudgetService.get_team_budget_overview(
        db=db,
        manager_id=current_user.id,
        year=year,
        month=month,
    )

    return overview
